File: C++ test files/QUESTIONS/LEETCODE/linklist_cycle.cpp
#include<bits/stdc++.h>
using namespace std;
struct node
{
    int data;
    node* next;
}*head;

node* createNode(int n)
{
    node* ptr = new node;
    ptr->data = n;
    ptr->next = NULL;
    return ptr;
}

node* juncNode(node* ptr)
{
    node *sptr, *fptr;
    sptr = fptr = ptr;

    while(fptr && fptr->next)
    {
        sptr = sptr->next;
        fptr = fptr->next->next;

        if(sptr == fptr)
        {
            sptr = ptr;
            while(sptr != fptr)
            {
                sptr = sptr->next;
                fptr = fptr->next;
            }
            return sptr;
        }
    }

    return NULL;
}

void printCycle(node* ptr)
{
    node *temp = ptr;
    do
    {
       cout<<temp->data<<" ";
        temp = temp->next; 

    }while (temp != ptr);
    cout<<endl;
}

int main()
{
    head = NULL;
    head = createNode(1);
    head = head;
    head->next = createNode(2);
    head->next->next = createNode(3);
    head->next->next->next = createNode(4);
    head->next->next->next->next = createNode(5);
    head->next->next->next->next->next = createNode(6);
    head->next->next->next->next->next->next = head->next->next;

    node *start = juncNode(head);

    if(!start)
        cout<<"no cycle"<<endl;
    else
    {
        printCycle(start);
    }
}

File: Java test files/linklist/LinklistCycle.java
package linklist;

public class LinklistCycle {

    static Node head;

    public static Node insert(int data){
        Node toAdd = new Node(data);
        toAdd.next = null;

        if(head == null) head = toAdd;

        else{
            Node temp = head;
            while(temp.next != null) 
                temp = temp.next;

            temp.next = toAdd;
        }

        return toAdd;
    }

    public static void print(){
        Node temp = head.next;
        while(temp != null){
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    static class Node{
        int data;
        Node next;

        public Node(int data){
            this.data = data;
            next = null;
        }
    }

    public static Node checkCycle() {
        Node slow_ptr = head, fast_ptr = head;
        while(fast_ptr != null && fast_ptr.next != null){
            slow_ptr = slow_ptr.next;
            fast_ptr = fast_ptr.next.next;

            if(slow_ptr == fast_ptr){
                slow_ptr = head;
                while(slow_ptr != fast_ptr){
                    slow_ptr = slow_ptr.next;
                    fast_ptr = fast_ptr.next;
                }
                return slow_ptr;
            }
        }

        return null;
    }

    public static void printCycle(Node juncNode) {
        Node temp = juncNode;
        do{
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        while(temp != juncNode);
        System.out.println();
    }
    public static void main(String[] args) {
        head = insert(0);
        head.next = insert(1);
        head.next.next = insert(2);
        head.next.next.next = insert(3);
        head.next.next.next.next = insert(4);
        head.next.next.next.next.next = insert(5);
        head.next.next.next.next.next.next = insert(6);
        head.next.next.next.next.next.next.next = head.next.next.next;
        
        Node juncNode = checkCycle();
        if(juncNode != null)
            printCycle(juncNode);
        else
            System.out.println("No cycle");  
    }
}
